<!DOCTYPE html>
<html>
<p><head>
<title>CLIPcontext - G2T Mode Report</title>
</head></p>
<p><img src="img/logo.png" alt="cc_logo"
    title="cc_logo" width="700" /></p>
<p><body style="font-family:sans-serif" link="#4b8cb6" vlink="#4b8cb6" alink="#4b8cb6"></p>
<h1>G2T context set comparison report</h1>
<p>List of available statistics to compare sites containing genomic context
with sites containing transcript context, generated by clipcontext g2t:</p>
<ul>
<li><a href="#set-stats">Context set statistics</a></li>
<li><a href="#ent-plot">Sequence complexity distribution</a></li>
<li><a href="#dint-plot">Di-nucleotide distribution</a></li>
<li><a href="#kmer-stats">Top k-mer statistics</a></li>
<li><a href="#gbt-stats">Target gene biotype statistics</a></li>
<li><a href="#so-stats">Site overlap statistics</a>
&nbsp;</li>
</ul>
<h2 id="set-stats">Context set statistics</h2>
<p><strong>Table:</strong> Context set statistics regarding sequence lengths
(min, max, mean, and median length) in nucleotides (nt),
sequence complexity (mean Shannon entropy over all sequences in the set)
and nucleotide contents (A, C, G, U).</p>
<table>
<thead>
<tr>
<th align="center">Attribute</th>
<th align="center">&nbsp; Transcript context &nbsp;</th>
<th align="center">&nbsp; Genomic context &nbsp;</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"># sites</td>
<td align="center">12667</td>
<td align="center">12667</td>
</tr>
<tr>
<td align="center">min site length</td>
<td align="center">56</td>
<td align="center">101</td>
</tr>
<tr>
<td align="center">max site length</td>
<td align="center">101</td>
<td align="center">101</td>
</tr>
<tr>
<td align="center">mean site length</td>
<td align="center">100.9</td>
<td align="center">101.0</td>
</tr>
<tr>
<td align="center">median site length</td>
<td align="center">101</td>
<td align="center">101</td>
</tr>
<tr>
<td align="center">mean complexity</td>
<td align="center">0.976</td>
<td align="center">0.973</td>
</tr>
<tr>
<td align="center">%A</td>
<td align="center">27.94</td>
<td align="center">26.86</td>
</tr>
<tr>
<td align="center">%C</td>
<td align="center">22.57</td>
<td align="center">22.22</td>
</tr>
<tr>
<td align="center">%G</td>
<td align="center">24.72</td>
<td align="center">24.16</td>
</tr>
<tr>
<td align="center">%U</td>
<td align="center">24.78</td>
<td align="center">26.77</td>
</tr>
</tbody>
</table>
<p>&nbsp;
&nbsp;</p>
<h2 id="ent-plot">Sequence complexity distribution</h2>
<p>The Shannon entropy is calculated for each sequence to measure
its information content (i.e., its complexity). A sequence with
equal amounts of all four nucleotides has an entropy value of 1.0
(highest possible). A sequence with equal amounts of two nucleotides
has an entropy value of 0.5. Finally, the lowest possible entropy is
achieved by a sequence which contains only one type of nucleotide.
Find the formula used to compute Shannon's entropy
<a href="https://www.ncbi.nlm.nih.gov/pubmed/15215465">here</a> (see CE formula).</p>
<p><img src="img/sequence_complexity_plot.png" alt="Sequence complexity distribution"
title="Sequence complexity distribution" width="500" /></p>
<p><strong>Figure:</strong> Sequence complexity (Shannon entropy
computed for each sequence) distributions for the genomic and transcript
context set.</p>
<p>&nbsp;</p>
<h2 id="dint-plot">Di-nucleotide distribution</h2>
<p>Di-nucleotide percentages are shown for the genomic and transcript
context dataset.</p>
<p><img src="img/dint_percentages_plot.png" alt="Di-nucleotide distribution"
title="Di-nucleotide distribution" width="1000" /></p>
<p><strong>Figure:</strong> Di-nucleotide percentages for the the genomic and transcript context dataset.</p>
<p>&nbsp;</p>
<h2 id="kmer-stats">Top k-mer statistics</h2>
<p><strong>Table:</strong> Top 10 3-mers for the genomic and transcript context set and their percentages in the respective sequence set. In case of uniform distribution with all 3-mers present, each 3-mer would have a percentage = 1.5625.</p>
<table>
<thead>
<tr>
<th align="center">Rank</th>
<th align="center">&nbsp; &nbsp; Transcript context &nbsp; &nbsp;</th>
<th align="center">&nbsp; &nbsp; Genomic context &nbsp; &nbsp;</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">1</td>
<td align="center">AGA (2.81 %)</td>
<td align="center">UUU (2.84 %)</td>
</tr>
<tr>
<td align="center">2</td>
<td align="center">CAG (2.67 %)</td>
<td align="center">CAG (2.59 %)</td>
</tr>
<tr>
<td align="center">3</td>
<td align="center">GAA (2.64 %)</td>
<td align="center">AGA (2.52 %)</td>
</tr>
<tr>
<td align="center">4</td>
<td align="center">CUG (2.52 %)</td>
<td align="center">CUG (2.46 %)</td>
</tr>
<tr>
<td align="center">5</td>
<td align="center">AAG (2.51 %)</td>
<td align="center">GAA (2.34 %)</td>
</tr>
<tr>
<td align="center">6</td>
<td align="center">AAA (2.46 %)</td>
<td align="center">AAG (2.33 %)</td>
</tr>
<tr>
<td align="center">7</td>
<td align="center">UGA (2.41 %)</td>
<td align="center">AAA (2.30 %)</td>
</tr>
<tr>
<td align="center">8</td>
<td align="center">UGG (2.32 %)</td>
<td align="center">UGA (2.30 %)</td>
</tr>
<tr>
<td align="center">9</td>
<td align="center">GGA (2.27 %)</td>
<td align="center">UGG (2.21 %)</td>
</tr>
<tr>
<td align="center">10</td>
<td align="center">AUG (2.19 %)</td>
<td align="center">AUG (2.04 %)</td>
</tr>
<tr>
<td align="center">...</td>
<td align="center">&nbsp;</td>
<td align="center">&nbsp;</td>
</tr>
<tr>
<td align="center"># distinct k-mers</td>
<td align="center">64 (100.00  %)</td>
<td align="center">64 (100.00  %)</td>
</tr>
</tbody>
</table>
<p>&nbsp;</p>
<p><strong>Table:</strong> Top 10 4-mers for the genomic and transcript context set and their percentages in the respective sequence set. In case of uniform distribution with all 4-mers present, each 4-mer would have a percentage = 0.390625.</p>
<table>
<thead>
<tr>
<th align="center">Rank</th>
<th align="center">&nbsp; &nbsp; Transcript context &nbsp; &nbsp;</th>
<th align="center">&nbsp; &nbsp; Genomic context &nbsp; &nbsp;</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">1</td>
<td align="center">AGAA (0.990 %)</td>
<td align="center">UUUU (1.010 %)</td>
</tr>
<tr>
<td align="center">2</td>
<td align="center">AAGA (0.976 %)</td>
<td align="center">AGAA (0.874 %)</td>
</tr>
<tr>
<td align="center">3</td>
<td align="center">GAAG (0.969 %)</td>
<td align="center">AAGA (0.843 %)</td>
</tr>
<tr>
<td align="center">4</td>
<td align="center">UGGA (0.879 %)</td>
<td align="center">GAAG (0.836 %)</td>
</tr>
<tr>
<td align="center">5</td>
<td align="center">CAGA (0.844 %)</td>
<td align="center">UGGA (0.772 %)</td>
</tr>
<tr>
<td align="center">6</td>
<td align="center">UGAA (0.809 %)</td>
<td align="center">CAGA (0.764 %)</td>
</tr>
<tr>
<td align="center">7</td>
<td align="center">AGGA (0.779 %)</td>
<td align="center">UGAA (0.730 %)</td>
</tr>
<tr>
<td align="center">8</td>
<td align="center">GCUG (0.772 %)</td>
<td align="center">GCUG (0.722 %)</td>
</tr>
<tr>
<td align="center">9</td>
<td align="center">GAAA (0.766 %)</td>
<td align="center">UUUG (0.701 %)</td>
</tr>
<tr>
<td align="center">10</td>
<td align="center">AAAA (0.746 %)</td>
<td align="center">AAAA (0.700 %)</td>
</tr>
<tr>
<td align="center">...</td>
<td align="center">&nbsp;</td>
<td align="center">&nbsp;</td>
</tr>
<tr>
<td align="center"># distinct k-mers</td>
<td align="center">256 (100.00  %)</td>
<td align="center">256 (100.00  %)</td>
</tr>
</tbody>
</table>
<p>&nbsp;</p>
<p><strong>Table:</strong> Top 10 5-mers for the genomic and transcript context set and their percentages in the respective sequence set. In case of uniform distribution with all 5-mers present, each 5-mer would have a percentage = 0.09765625.</p>
<table>
<thead>
<tr>
<th align="center">Rank</th>
<th align="center">&nbsp; &nbsp; Transcript context &nbsp; &nbsp;</th>
<th align="center">&nbsp; &nbsp; Genomic context &nbsp; &nbsp;</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">1</td>
<td align="center">GAAGA (0.4129 %)</td>
<td align="center">UUUUU (0.3674 %)</td>
</tr>
<tr>
<td align="center">2</td>
<td align="center">AAGAA (0.3679 %)</td>
<td align="center">GAAGA (0.3458 %)</td>
</tr>
<tr>
<td align="center">3</td>
<td align="center">AGAAG (0.3592 %)</td>
<td align="center">AAGAA (0.3186 %)</td>
</tr>
<tr>
<td align="center">4</td>
<td align="center">UGAAG (0.3242 %)</td>
<td align="center">AGAAG (0.3085 %)</td>
</tr>
<tr>
<td align="center">5</td>
<td align="center">AGAAA (0.2946 %)</td>
<td align="center">UGAAG (0.2828 %)</td>
</tr>
<tr>
<td align="center">6</td>
<td align="center">UGCUG (0.2936 %)</td>
<td align="center">UGCUG (0.2700 %)</td>
</tr>
<tr>
<td align="center">7</td>
<td align="center">UGGAG (0.2890 %)</td>
<td align="center">AGAAA (0.2608 %)</td>
</tr>
<tr>
<td align="center">8</td>
<td align="center">GAGGA (0.2765 %)</td>
<td align="center">UGGAG (0.2527 %)</td>
</tr>
<tr>
<td align="center">9</td>
<td align="center">GAUGA (0.2740 %)</td>
<td align="center">AUUUU (0.2494 %)</td>
</tr>
<tr>
<td align="center">10</td>
<td align="center">AAAGA (0.2704 %)</td>
<td align="center">UGCAG (0.2457 %)</td>
</tr>
<tr>
<td align="center">...</td>
<td align="center">&nbsp;</td>
<td align="center">&nbsp;</td>
</tr>
<tr>
<td align="center"># distinct k-mers</td>
<td align="center">1024 (100.00  %)</td>
<td align="center">1024 (100.00  %)</td>
</tr>
</tbody>
</table>
<p>&nbsp;
&nbsp;</p>
<h2 id="gbt-stats">Target gene biotype statistics</h2>
<p><strong>Table:</strong> Target gene biotype counts for transcript context sites and their percentages
(count normalized by total count for the respective gene biotype).</p>
<table>
<thead>
<tr>
<th align="center">&nbsp; Gene biotype &nbsp;</th>
<th align="center">&nbsp; Target count &nbsp;</th>
<th align="center">&nbsp; Total count &nbsp;</th>
<th align="center">&nbsp; Percentage &nbsp;</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">protein_coding</td>
<td align="center">2562</td>
<td align="center">19968</td>
<td align="center">12.83 %</td>
</tr>
<tr>
<td align="center">snoRNA</td>
<td align="center">10</td>
<td align="center">942</td>
<td align="center">1.06 %</td>
</tr>
<tr>
<td align="center">misc_RNA</td>
<td align="center">6</td>
<td align="center">2220</td>
<td align="center">0.27 %</td>
</tr>
<tr>
<td align="center">lncRNA</td>
<td align="center">6</td>
<td align="center">16880</td>
<td align="center">0.04 %</td>
</tr>
<tr>
<td align="center">rRNA_pseudogene</td>
<td align="center">5</td>
<td align="center">499</td>
<td align="center">1.00 %</td>
</tr>
<tr>
<td align="center">polymorphic_pseudogene</td>
<td align="center">2</td>
<td align="center">55</td>
<td align="center">3.64 %</td>
</tr>
<tr>
<td align="center">miRNA</td>
<td align="center">1</td>
<td align="center">1879</td>
<td align="center">0.05 %</td>
</tr>
<tr>
<td align="center">unprocessed_pseudogene</td>
<td align="center">1</td>
<td align="center">2627</td>
<td align="center">0.04 %</td>
</tr>
<tr>
<td align="center">scaRNA</td>
<td align="center">1</td>
<td align="center">49</td>
<td align="center">2.04 %</td>
</tr>
</tbody>
</table>
<p>&nbsp;
&nbsp;</p>
<h2 id="so-stats">Site overlap statistics</h2>
<p><strong>Table:</strong> Site overlap statistics, showing the top 20 targeted
regions (transcripts and corresponding genes), with the # overlaps == # of
transcript context sites overlapping with the region.</p>
<table>
<thead>
<tr>
<th align="center">&nbsp; # overlaps &nbsp;</th>
<th align="center">&nbsp; Transcript ID &nbsp;</th>
<th align="center">&nbsp; &nbsp; Transcript biotype &nbsp; &nbsp;</th>
<th align="center">&nbsp; Gene ID &nbsp;</th>
<th align="center">&nbsp; Gene name &nbsp;</th>
<th align="center">&nbsp; &nbsp; Gene biotype &nbsp; &nbsp;</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">120</td>
<td align="center">ENST00000375254</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000127481</td>
<td align="center">UBR4</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">75</td>
<td align="center">ENST00000314191</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000253729</td>
<td align="center">PRKDC</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">65</td>
<td align="center">ENST00000369850</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000196924</td>
<td align="center">FLNA</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">58</td>
<td align="center">ENST00000262854</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000086758</td>
<td align="center">HUWE1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">55</td>
<td align="center">ENST00000269122</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000141367</td>
<td align="center">CLTC</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">55</td>
<td align="center">ENST00000582970</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000173821</td>
<td align="center">RNF213</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">53</td>
<td align="center">ENST00000261609</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000128731</td>
<td align="center">HERC2</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">49</td>
<td align="center">ENST00000304992</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000174231</td>
<td align="center">PRPF8</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">47</td>
<td align="center">ENST00000446231</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000157106</td>
<td align="center">SMG1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">46</td>
<td align="center">ENST00000323853</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000144028</td>
<td align="center">SNRNP200</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">46</td>
<td align="center">ENST00000398571</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000115464</td>
<td align="center">USP34</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">45</td>
<td align="center">ENST00000378308</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000124486</td>
<td align="center">USP9X</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">44</td>
<td align="center">ENST00000377560</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000173064</td>
<td align="center">HECTD4</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">43</td>
<td align="center">ENST00000360184</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000197102</td>
<td align="center">DYNC1H1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">43</td>
<td align="center">ENST00000314888</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000137076</td>
<td align="center">TLN1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">41</td>
<td align="center">ENST00000358273</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000196712</td>
<td align="center">NF1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">40</td>
<td align="center">ENST00000317147</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000125107</td>
<td align="center">CNOT1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">39</td>
<td align="center">ENST00000325083</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000078674</td>
<td align="center">PCM1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">38</td>
<td align="center">ENST00000342981</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000114867</td>
<td align="center">EIF4G1</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">37</td>
<td align="center">ENST00000421745</td>
<td align="center">protein_coding</td>
<td align="center">ENSG00000115760</td>
<td align="center">BIRC6</td>
<td align="center">protein_coding</td>
</tr>
<tr>
<td align="center">...</td>
<td align="center">&nbsp;</td>
<td align="center">&nbsp;</td>
<td align="center">&nbsp;</td>
<td align="center">&nbsp;</td>
<td align="center">&nbsp;</td>
</tr>
</tbody>
</table>
<p>&nbsp;
&nbsp;</p>
</body>
</html> 
